### PR TITLE
Dynamo Core version should not be replaced after customer manually installed a later D4R version

### PR DESCRIPTION
**Pull request by ZiyunShang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

If Dynamo as an Addin in Revit, Improve DynamCorePath so it can be controlled more precisely.

### Declarations

Check these if you believe they are true

- [ ] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] Snapshot of UI changes, if any.

### Reviewers

@AndyDu1985 

### FYIs

@QilongTang @mjkkirschner

### CODE REVIEW

File: src/DynamoRevit/DynamoRevitApp.cs [hunk 1]
```diff
@@ -381,8 +381,41 @@ public static bool DynamoButtonEnabled
...
+        private static Boolean IsRevitAddin(UIControlledApplication application)
+        {
+            if (application == null)
+                return false;
+            var revitVersion = application.ControlledApplication.VersionNumber;
+            if (int.Parse(revitVersion) < 2020)
+               return false;
+            var dynamoRevitRoot = Path.GetDirectoryName(Path.GetDirectoryName(assemblyName));
+            var RevitRoot = Path.GetDirectoryName(application.GetType().Assembly.Location);
+            if (dynamoRevitRoot.Contains(RevitRoot))
```
Comment (AndyDu1985) [inline src/DynamoRevit/DynamoRevitApp.cs]: Is there a string.startWith(...)?

File: src/DynamoRevit/DynamoRevitApp.cs [hunk 2]
```diff
@@ -381,8 +381,41 @@ public static bool DynamoButtonEnabled
...
         }
 
-        private bool TryResolveDynamoCore()
+        /// <summary>
+        /// Whether the DynamoCore and DynamoRevit are in Revit's Addin, 
+        /// and Revit version is more than 2020
+        /// </summary>
+        /// <param name="application"></param>
+        /// <returns>True is that Dynamo and DynamoRevit are in Revit Addins</returns>
+        private static Boolean IsRevitAddin(UIControlledApplication application)
```
Comment (AndyDu1985) [inline src/DynamoRevit/DynamoRevitApp.cs]: Can it rename to IsRevitInternalAddin?

File: src/DynamoRevit/DynamoRevitApp.cs (hunk 1 above)
Comment (ZiyunShang) [reply src/DynamoRevit/DynamoRevitApp.cs]: I have changed this method.

File: src/DynamoRevit/DynamoRevitApp.cs [hunk 3]
```diff
@@ -381,8 +381,41 @@ public static bool DynamoButtonEnabled
             set;
         }
 
-        private bool TryResolveDynamoCore()
+        /// <summary>
+        /// Whether the DynamoCore and DynamoRevit are in Revit's Addin,
```
Comment (mjkkirschner) [inline src/DynamoRevit/DynamoRevitApp.cs]: to you mean to say `in Revit's Addin folder`?

File: src/DynamoRevit/DynamoRevitApp.cs [hunk 4]
```diff
@@ -381,8 +381,41 @@ public static bool DynamoButtonEnabled
...
+                       return true;
+                }
+            }
+            return false;
+        }
+
+        private bool TryResolveDynamoCore(UIControlledApplication application)
         {
+            if(IsRevitInternalAddin(application))
+                dynamopath= Path.GetDirectoryName(Path.GetDirectoryName(assemblyName));
```
Comment (mjkkirschner) [inline src/DynamoRevit/DynamoRevitApp.cs]: just a style thing but I think this repo usually prefers braces when possible.

File: src/DynamoRevit/DynamoRevitApp.cs (hunk 3 above)
Comment (ZiyunShang) [reply src/DynamoRevit/DynamoRevitApp.cs]: Yes, it should be in Revit's internal addin folder.

File: src/DynamoRevit/DynamoRevitApp.cs (hunk 4 above)
Comment (ZiyunShang) [reply src/DynamoRevit/DynamoRevitApp.cs]: I will add braces here.

**Comment by mjkkirschner** [comment]
Hi @ZiyunShang - thanks for this work - can you give a brief overview of what this change means going forward - DynamoRevit will always load the dynamoCore from the Revit addin folder? What about if a dev manually specifies another core using the config - will it be used instead?

 Or the proper testing setup would be to point a custom addin at the dev's local DynamoRevit and Dynamo bin folders?

**Comment by mjkkirschner** [comment]
@ZiyunShang thanks! - we may want to add these instructions to the readme of this repo - I'll send a PR and ping you for review if thats ok?

**Pull request merged**